ingestion-core/src/main/java/com/mozilla/telemetry/ingestion/core/transform/PubsubMessageToObjectNode.java

```diff
@@ -595,6 +642,59 @@ private String getAndCacheBqName(String name) {
       }
     }
 
+    /**
+     * Encode the given JSON object as a compact string.
+     *
+     * <p>We maintain a BigQuery persistent UDF that can parse a variety of formats, so any format
+     * used here must first be supported in:
+     * https://github.com/mozilla/bigquery-etl/blob/master/mozfun/hist/extract/udf.sql
+     *
+     * <p>Schema validation should generally ensure that all histograms are well-formed, but we
+     * perform some light sanity checking on the values and fall back to encoding the histogram
+     * as a JSON string if anything seems off.
+     */
+    private TextNode compactHistogramEncoding(JsonNode o, String fieldName) {
+      TextNode jsonEncoding = TextNode.valueOf(Json.asString(o));
+      int histogramType = o.path("histogram_type").asInt(-1);
+      if (histogramType < 0 || histogramType > 5) {
+        incrementInvalidHistogramType();
+        return jsonEncoding;
+      }
+      long sum = o.path("sum").asLong(-1);
+      if (sum < 0) {
+        incrementInvalidHistogramSum();
+        return jsonEncoding;
+      }
+      JsonNode values = o.path("values");
+      if (histogramType == 4) {
+        // Type 4 histograms are "count" histograms representing only a single value.
+        return TextNode.valueOf(Long.toString(sum));
+      } else if (histogramType == 2 && fieldName.startsWith("use_counter2")) {
+        // "use counters" represent the bulk of histogram data sent in main pings, so we provide
+        // additional optimization for this case; although they are encoded as "boolean" histograms
+        // (type 2), the "0" (false) bucket is never populated, meaning they only encode a single
+        // value.
+        return TextNode.valueOf(Long.toString(sum));
+      } else if (histogramType == 2) {
+        // Type 2 are "boolean" histograms where bucket "0" is a count of false values and
+        // bucket "1" is a count of true values.
+        return TextNode.valueOf(
+            String.format("%d,%d", values.path("0").longValue(), values.path("1").longValue()));
+      } else {
+        int bucketCount = o.path("bucket_count").asInt(-1);
+        long rangeLo = o.path("range").path(0).asLong(-1);
+        long rangeHi = o.path("range").path(1).asLong(-1);
+        String valString = Json.asString(values).replace("{", "").replace("}", "").replace("\"",
+            "");
+        if (bucketCount <= 0 || rangeLo < 0 || rangeHi < 0) {
+          incrementInvalidHistogramRange();
+          return jsonEncoding;
+        }
+        return TextNode.valueOf(String.format("%d;%d;%d;%d,%d;%s", //
+            bucketCount, histogramType, sum, rangeLo, rangeHi, valString));
```

Comment:
I think we should put `valString` first in this list, rather than last, so that the position in the semicolon separated list consistent.

It's fine as-is if it's too late for this change because we don't want to deviate from the proposal.